### PR TITLE
msg/async: use faster clear method to delete containers

### PR DESCRIPTION
**Pull request by batrick**
Erasing each element is unnecessarily slower than just doing a clear after any
necessary iteration.

Signed-off-by: Patrick Donnelly <pdonnell@redhat.com>

### CODE REVIEW

**Comment by batrick** [comment]
From #26348

**Comment by batrick** [comment]
retest this please

**Comment by batrick** [comment]
retest this please

**Comment by xiexingguo** [comment]
@tchaikov @badone  Saw this multiple times...
```
/home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/test/librbd/object_map/test_mock_LockRequest.cc:61:3: internal compiler error: Segmentation fault
   }
   ^
Please submit a full bug report,
with preprocessed source if appropriate.
See <http://bugzilla.redhat.com/bugzilla> for instructions.
Preprocessed source stored into /tmp/cchJZk4J.out file, please attach this to your bugreport.
make[3]: *** [src/test/librbd/CMakeFiles/unittest_librbd.dir/object_map/test_mock_LockRequest.cc.o] Error 1
make[2]: *** [src/test/librbd/CMakeFiles/unittest_librbd.dir/all] Error 2
make[2]: *** Waiting for unfinished jobs....
[100%] Built target radosgw
[100%] Built target ceph-dencoder
make[1]: *** [CMakeFiles/tests.dir/rule] Error 2
make: *** [tests] Error 2
```

**Comment by tchaikov** [comment]
retest this please.

File: src/msg/async/AsyncMessenger.cc [hunk 1]
```diff
@@ -931,18 +927,18 @@ int AsyncMessenger::reap_dead()
...
-    ldout(cct, 5) << __func__ << " delete " << p << dendl;
-    auto conns_it = conns.find(*p->peer_addrs);
-    if (conns_it != conns.end() && conns_it->second == p)
-      conns.erase(conns_it);
-    accepting_conns.erase(p);
-    deleted_conns.erase(it);
-    ++num;
+
+  {
+    Mutex::Locker l2(deleted_lock);
```
Comment (tchaikov) [inline src/msg/async/AsyncMessenger.cc]: could instead use `std::scoped_lock` , like
```c++
std::scoped_lock locker{lock, deleted_lock};
```

**Comment by badone** [comment]
> @tchaikov @badone Saw this multiple times...
> 
> ```
> /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/test/librbd/object_map/test_mock_LockRequest.cc:61:3: internal compiler error: Segmentation fault
>    }
>    ^
> Please submit a full bug report,
> with preprocessed source if appropriate.
> See <http://bugzilla.redhat.com/bugzilla> for instructions.
> Preprocessed source stored into /tmp/cchJZk4J.out file, please attach this to your bugreport.
> make[3]: *** [src/test/librbd/CMakeFiles/unittest_librbd.dir/object_map/test_mock_LockRequest.cc.o] Error 1
> make[2]: *** [src/test/librbd/CMakeFiles/unittest_librbd.dir/all] Error 2
> make[2]: *** Waiting for unfinished jobs....
> [100%] Built target radosgw
> [100%] Built target ceph-dencoder
> make[1]: *** [CMakeFiles/tests.dir/rule] Error 2
> make: *** [tests] Error 2
> ```

@xiexingguo @tchaikov Was this using gcc8 (devtoolset-8)?

**Comment by xiexingguo** [comment]
@badone Not being able to see it again (so far) after https://github.com/ceph/ceph/pull/27333

**Comment by xiexingguo** [comment]
> @xiexingguo @tchaikov Was this using gcc8 (devtoolset-8)?

I don't think so. The retest command is issued before https://github.com/ceph/ceph/pull/27333 gets merged.

**Comment by batrick** [comment]
rebased

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-04-07_01:07:57-rados-wip-kefu-testing-2019-04-05-1411-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-04-08_08:52:22-rados-wip-kefu-testing-2019-04-05-1411-distro-basic-smithi/#

**Pull request merged**